File: projects/agentic-automation-framework/src/playwright/runner.js
import { chromium } from "playwright";
import fs from "fs";
import path from "path";

export async function runWorkflow(workflow, { headless = true } = {}) {
  const browser = await chromium.launch({ headless });
  const context = await browser.newContext();
  const page = await context.newPage();
  const started = Date.now();

  try {
    await page.goto(workflow.url, { waitUntil: "domcontentloaded" });

    for (const step of workflow.steps || []) {
      switch (step.action) {
        case "click":
          await page.locator(step.selector).click();
          break;
        case "type":
          await page.locator(step.selector).fill(step.text ?? "");
          break;
        case "expectUrlContains": {
          const url = page.url();
          if (!url.includes(step.value)) throw new Error(`URL assertion failed. Expected contains '${step.value}', got '${url}'`);
          break;
        }
        case "waitFor":
          await page.locator(step.selector).waitFor({ state: step.state || "visible", timeout: step.timeout || 10000 });
          break;
        case "screenshot": {
          const out = path.resolve(step.path || "reports/screenshot.png");
          fs.mkdirSync(path.dirname(out), { recursive: true });
          await page.screenshot({ path: out, fullPage: true });
          break;
        }
        default:
          throw new Error(`Unsupported action: ${step.action}`);
      }
    }

    return {
      ok: true,
      name: workflow.name || "workflow",
      durationMs: Date.now() - started,
      finalUrl: page.url()
    };
  } catch (e) {
    return {
      ok: false,
      name: workflow.name || "workflow",
      durationMs: Date.now() - started,
      error: String(e)
    };
  } finally {
    await context.close();
    await browser.close();
  }
}

File: projects/testops-platform/app/mobile/runner.py
from pathlib import Path
from datetime import datetime, UTC
import json

REPORT = Path("reports/mobile-report.json")

DEVICES = [
    {"name": "iPhone 13", "viewport": {"width": 390, "height": 844}, "user_agent": "Mobile Safari"},
    {"name": "Pixel 7", "viewport": {"width": 412, "height": 915}, "user_agent": "Chrome Mobile"},
    {"name": "iPad Mini", "viewport": {"width": 768, "height": 1024}, "user_agent": "Mobile Safari"},
]


def list_devices():
    return DEVICES


def run_mobile_checks(url: str = "https://example.com", device: str = "iPhone 13", simulate: bool = True):
    selected = next((d for d in DEVICES if d["name"] == device), DEVICES[0])

    if simulate:
        result = {
            "ok": True,
            "mode": "simulate",
            "device": selected,
            "url": url,
            "checks": [
                {"id": "viewport_render", "status": "PASS"},
                {"id": "responsive_layout", "status": "PASS"},
                {"id": "touch_target_smoke", "status": "PASS"},
            ],
        }
    else:
        # Optional real execution path
        from playwright.sync_api import sync_playwright

        with sync_playwright() as p:
            browser = p.chromium.launch(headless=True)
            context = browser.new_context(viewport=selected["viewport"], user_agent=selected["user_agent"])
            page = context.new_page()
            page.goto(url, wait_until="domcontentloaded")
            shot = Path("reports") / f"mobile-{device.replace(' ', '-').lower()}.png"
            shot.parent.mkdir(parents=True, exist_ok=True)
            page.screenshot(path=str(shot), full_page=True)
            browser.close()

        result = {
            "ok": True,
            "mode": "real",
            "device": selected,
            "url": url,
            "screenshot": str(shot),
            "checks": [
                {"id": "viewport_render", "status": "PASS"},
                {"id": "responsive_layout", "status": "PASS"},
                {"id": "touch_target_smoke", "status": "PASS"},
            ],
        }

    payload = {
        "ts": datetime.now(UTC).isoformat(),
        **result,
    }
    REPORT.parent.mkdir(parents=True, exist_ok=True)
    REPORT.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    return payload


def last_report():
    if not REPORT.exists():
        return {"ok": False, "error": "no mobile report yet"}
    try:
        return json.loads(REPORT.read_text())
    except Exception:
        return {"ok": False, "error": "failed to read mobile report"}
